feat: [sc-58279] [core] add `tiledb_array_schema_get_enumeration` API

rroelke opened the pull request:
Story details: https://app.shortcut.com/tiledb-inc/story/58279

> In tiledb-rs we have have "property-based testing" strategies which generate arbitrary tiledb schema. We would like to extend these strategies to generate arbitrary enumerations to put in those schemata.
> 
> As a step towards this we must add an API tiledb_array_schema_get_enumeration so that we can load the values in an enumeration without having a pointer to the open array in the same function.
> 
> It is not required for this API to actually load the enumeration, just to return its contents if it is already loaded. There should be a graceful signal to distinguish between an enumeration not being found and an enumeration not being loaded.

This pull request goes all the way and adds functions `tiledb_array_schema_get_enumeration_from_name` and `tiledb_array_schema_get_enumeration_from_attribute_name` which load the contents of an enumeration and return a handle to the user.

See [this comment](https://github.com/TileDB-Inc/TileDB/pull/5359#issuecomment-2448578815) for further discussion.

---
TYPE: FEATURE | C_API | CPP_API
DESC: add tiledb_array_schema_get_enumeration

ypatia commented on tiledb/api/c_api/array_schema/array_schema_api_experimental.h:
```diff
@@ -67,6 +67,36 @@ TILEDB_EXPORT capi_return_t tiledb_array_schema_timestamp_range(
+ * @param[out] enmr The enumeration object to retrieve.
+ * @return `TILEDB_OK` for success and `TILEDB_ERR` for error.
+ */
+TILEDB_EXPORT capi_return_t tiledb_array_schema_get_enumeration(
```
I think that the different behavior we are introducing here between `tiledb_array_get_enumeration` and `tiledb_array_schema_get_enumeration` APIs will be confusing to the user (as it was to me while reviewing) , ie that we expect form the former to load the enumeration if not loaded and from the latter not even though the syntax is so similar.
If we need one that doesn't load the enumeration I'd make this one explicit, eg. `tiledb_array_schema_get_loaded_enumeration` , and let `tiledb_array_schema_get_enumeration` (if added) have the same behavior as `tiledb_array_get_enumeration`.

rroelke replied:
Took a peek and I see that we'd have to get a ClientResources something or other into ArraySchema (which @davisp did warn me about), doesn't seem worth it at the moment so we're going with the name change to `tiledb_array_schema_get_enumeration_if_loaded`.  0bf188d27

ypatia commented on test/src/unit-cppapi-enumerations.cc:
```diff
@@ -313,6 +325,67 @@ TEST_CASE_METHOD(
+TEST_CASE_METHOD(
+    CPPEnumerationFx,
+    "CPP: Enumerations From Disk - ArraySchema::get_enumeration",
+    "[enumeration][array-schema-get-enumeration]") {
```
please also add the `[rest]` tag so that this runs on REST-CI as well and verifies the behavior is identical/right for `tiledb://` URIs.

rroelke replied:
5d5c3b

shaunrd0 reviewed (commented):
Just left some small comments, looks great so far :+1:

davisp commented on tiledb/sm/cpp_api/array_schema_experimental.h:
```diff
@@ -189,6 +189,31 @@ class ArraySchemaExperimental {
   }
 
+  /**
+   * Retrieve an enumeration from the array schema, if has already been loaded
```
Pretty sure you meant "if it has already".

rroelke replied:
c1efcf0

davisp reviewed (approved):
+1 from me. @ypatia already suggested the REST test which is all I had thought of.

rroelke commented on tiledb/api/c_api/array_schema/test/unit_capi_array_schema.cc:
```diff
@@ -361,6 +361,68 @@ TEST_CASE(
+        0,
+        &enumeration);
+    REQUIRE(tiledb_status(rc) == TILEDB_OK);
+    tiledb_array_schema_add_enumeration(x.ctx(), x.schema, enumeration);
```
dc2b5ae

rroelke commented on test/src/unit-cppapi-enumerations.cc:
```diff
@@ -44,6 +44,18 @@
 
 using namespace tiledb;
 
+static bool is_equivalent_enumeration(
```
0a77551704

rroelke commented on tiledb/sm/cpp_api/array.h:
```diff
@@ -708,12 +710,16 @@ class Array {
     tiledb_array_schema_t* schema;
-    ctx.handle_error(
-        tiledb_array_schema_load(ctx.ptr().get(), uri.c_str(), &schema));
+    ctx.handle_error(tiledb_array_schema_load_with_config(
```
After I did the overload it did occur to be that something like a function pointer would probably break.  Breaking a Cython declaration (for example) was something I didn't think of.

> maybe we could get the config from the Context so that both APIs load enumerations on schema load

The config is deliberately cleared here
```
  // Use a default constructed config to load the schema with default options.
  *array_schema = tiledb_array_schema_t::make_handle(
      load_array_schema(ctx->context(), sm::URI(array_uri), sm::Config()));
```
in commit 9b4e5ea0c8

was there a specific configuration option that we wanted to avoid here?  Is it your recommendation to change this instead to `ctx->context()->config()`?

shaunrd0 replied:
Good point, I'm not sure how destructive that would be on CI now that you mention it.. It could be that using the context configuration in `tiledb_array_schema_load` I ran into a lot of failures so I went for a totally separate `_with_config` API instead. I was just pointing out the potential breakage for tiledb-py and trying to think of a way around that. Feel free to give it a shot if you'd like, but if using the context configurations causes too much CI troubles I'd say this is fine too if it works for you

```c++
  static ArraySchema load_schema(
      const Context& ctx,
      const std::string& uri) {
    tiledb_array_schema_t* schema;
    ctx.handle_error(tiledb_array_schema_load(
        ctx.ptr().get(), uri.c_str(), &schema));
    return ArraySchema(ctx, schema);
  }

  static ArraySchema load_schema_with_config(
      const Context& ctx,
      const std::string& uri,
      const Config& config) {
    tiledb_array_schema_t* schema;
    ctx.handle_error(tiledb_array_schema_load_with_config(
        ctx.ptr().get(), config.ptr().get(), uri.c_str(), &schema));
    return ArraySchema(ctx, schema);
  }
```

rroelke replied:
Nah I'm taking the easy route here.  754f239e undoes the default argument and adds the separate named option as suggested.

rroelke commented:
> +1 from me. @ypatia already suggested the REST test which is all I had thought of.

@davisp I mis-clicked when I requested a re-review, but feel free :)

All the changes are strictly in response to the review comments.

shaunrd0 reviewed (approved):
Nice! :+1:

rroelke commented:
After using this branch in my `tiledb-rs` branch I was not very satisfied, so instead of debugging the REST failure today I instead spent time making the schema lookup actually load the enumeration (instead of returning data only if the enumeration was already loaded).

While doing so I also noticed that `Array::get_enumeration` uses the _attribute_ name as the lookup key, not the enumeration name. This does not match what I had done here.

So the most recent commit does the following:
- Removes `tiledb_array_schema_get_enumeration_if_loaded`
- Adds `tiledb_array_schema_get_enumeration_from_name`
- Adds `tiledb_array_schema_get_enumeration_from_attribute_name`
- Adds the internal method `ArraySchema::load_enumeration` which does the I/O and updates the `ArraySchema` internal pointer, and calls this method from the C API handler

I reckon this merits a re-review since it's a pretty big change.

rroelke commented on tiledb/api/c_api/array_schema/CMakeLists.txt:
```diff
@@ -36,6 +36,7 @@ commence(object_library capi_array_schema_stub)
   this_target_sources(${SOURCES})
   this_target_link_libraries(export)
   this_target_object_libraries(array_schema)
+  this_target_object_libraries(array)
```
This is because `tiledb/sm/array_schema/array_schema.cc` calls upon `tiledb/sm/array/array_directory.cc`. So now the unit test needs to link with that.

Shelnutt2 commented on tiledb/sm/array_schema/array_schema.cc:
```diff
@@ -1128,6 +1132,34 @@ const std::string& ArraySchema::get_enumeration_path_name(
   return iter->second;
 }
 
+void ArraySchema::load_enumeration(Context& ctx, const std::string& enmr_name) {
```
This function needs to support `remote_array` case like the `Array::load_enumerations`. I think we can reuse the same route and only minor adjustments will be needed in serialization functions to support this.

rroelke commented:
@shaunrd0 @davisp @ypatia 

I think this is all set, CI passes and I don't intend to make any more changes.

Since the last review I determined I was unsatisfied with the "return the enumeration only if loaded" approach and have updated it to load the enumeration as part of the API request.

The code was easy enough but the troubles come with avoiding circular dependencies.  `array/array_directory.cc` is used to do the I/O, but the `array` "object library" depends on `array_schema` - so we can't have `array_schema` circularly depend on `array`.  The solution to this is to have the C API handler call a separate function to load the enumeration.

The other change since the last review is tweaking the API to be `tiledb_array_schema_get_enumeration_from_name` and `tiledb_array_schema_get_enumeration_from_attribute_name`.  I noticed that `tiledb_array_get_enumeration` used the attribute name, not the enumeration name, as the lookup key, and I wanted to disambiguate for these versions.  The implementation of `tiledb_array_schema_get_enumeration_from_attribute_name` calls `tiledb_attribute_get_enumeration` which required a few more changes to some "object libraries" but I thought this would make sense so that any errors would be presented the same way.

The good news is that my downstream `tiledb-rs` is very happy with this code and does exactly what I want it to. Hooray!

shaunrd0 commented on tiledb/sm/array/array.h:
```diff
@@ -1214,6 +1215,17 @@ class Array {
+ * @param enmr_name the requested enumeration
+ * @param schema the target schema
+ */
+void load_enumeration_into_schema(
```
Could this maybe go into `array_schema_operations.cc`, or did you also run into problems there with dependencies? At a glance it seems like it could work

rroelke replied:
I believe it won't work there either.  I am pretty new to cmake but my impression is that all the `array_schema/*.cc` files are compiled into the same "object library", hence putting this code in `array_schema_operations.cc` would also have the circular dependency between `array` and `array_schema`

shaunrd0 commented on test/support/src/array_schema_helpers.cc:
```diff
@@ -0,0 +1,54 @@
+
+namespace tiledb::test {
+
+bool is_equivalent_enumeration(
```
Thanks for putting this here, I'm likely going to use it when you merge

shaunrd0 reviewed (approved):
Very nice! Most comments were nits on updating docs since things have changed a bit since the first pass. I don't think there should be too many changes from my comments but if you want me to take another look let me know. Just as a heads up we should probably update the PR description before merging.

rroelke commented on tiledb/sm/array_schema/array_schema_operations.cc:
```diff
@@ -238,12 +238,35 @@ shared_ptr<ArraySchema> load_array_schema(
+      auto tracker = ctx.resources().ephemeral_memory_tracker();
+      // Pass an empty list of enumeration names. REST will use timestamps to
+      // load all enumerations on all schemas for the array within that range.
+      auto ret = rest_client.post_enumerations_from_rest(
```
Yeah, I didn't love this implementation, but... it gets the job done without having any obvious problems

ypatia replied:
@shaunrd0 please add a TODO in your https://github.com/TileDB-Inc/TileDB/pull/5181 or its ticket to address this. I don't know of any other place where we do 2 REST requests so we should really treat that as temporary and fix it asap.

rroelke commented on tiledb/api/c_api/array_schema/array_schema_api.cc:
```diff
@@ -181,6 +182,54 @@ capi_return_t tiledb_array_schema_timestamp_range(
+  }
+
+  std::string enumeration_name(enumeration_name_inner->view());
+  return api_entry_with_context<
```
Surprisingly I think the answer is "no" without making changes elsewhere.

`tiledb_string_t` has no `c_str()`, just `view()`

and I was initially surprised to see that `string_view` *also* has no `c_str()`, though in retrospect it makes sense.

ypatia commented on tiledb/sm/array/array.cc:
```diff
@@ -2071,4 +2074,59 @@ void ensure_supported_schema_version_for_read(format_version_t version) {
+        tracker);
+
+    // response is a map {schema: [enumerations]}
+    // we should be the only schema, and expect only one enumeration
```
It'd be nice to add a REST test that evolves the array with an additional enumeration and then check if you get that new enumeration on the new array schema when you call `tiledb_array_schema_get_enumeration_from_name`.

rroelke replied:
Sure, I'll add something in a follow-up PR

rroelke replied:
https://github.com/TileDB-Inc/TileDB/pull/5368/commits/79c4eaf0ba9623e4c59e3f094f0b538093cb35f6

ypatia reviewed (commented):
I was sick Thursday and this went in before I could review it. I am sorry for the delayed comments, but the ones around the new C-API body would need to be addressed.

rroelke commented on tiledb/api/c_api/array_schema/array_schema_api.cc:
```diff
@@ -181,6 +182,54 @@ capi_return_t tiledb_array_schema_timestamp_range(
+  ensure_output_pointer_is_valid(enumeration);
+
+  tiledb_attribute_t* attribute;
+  capi_return_t getattr = tiledb_array_schema_get_attribute_from_name(
```
I chose to do this so that if the attribute were missing we would get the same error message.  But if this is not something we are supposed to do then I'll change it.

As for the leak - oops!  Thanks for catching that

ypatia replied:
@rroelke Can we please fix that before it goes into released code? We are planning a 2.27 in the upcoming week. Thanks!

rroelke replied:
https://github.com/TileDB-Inc/TileDB/pull/5368/commits/bcaa3e5d621ef8500a6051a76645c7126297d556

rroelke commented on tiledb/sm/rest/rest_client_remote.cc:
```diff
@@ -220,10 +220,13 @@ RestClientRemote::get_array_schema_from_rest(const URI& uri) {
+  auto array_schema = serialization::array_schema_deserialize(
+      serialization_type_, returned_data, memory_tracker_);
+
+  array_schema->set_array_uri(uri);
```
Without this we get an error on the client side in `tiledb::sm::load_enumeration_into_schema`.

```
if (array_schema.array_uri().is_tiledb()) {
    /* it's remote, talk to REST server */
} else {
    /* it's local, read filesystem */
}
```
If we don't set the array URI here then it is blank. As a result `is_tiledb` is false and the schema tries to read the local filesystem, which results in error when it tries to open the empty path.
